qutip/nonmarkov/bofin.py

```diff
@@ -0,0 +1,1445 @@
+        promoted to a Liouvillian if it was a Hamiltonian).
+
+    options : :class:`qutip.solver.Options`
+        Generic solver options.
+        If set to None the default options will be used.
+    """
+    def __init__(
+        self, H_sys, coup_op, coup_strength, temperature,
+        N_cut, N_exp, cut_freq, bnd_cut_approx=False, options=None,
+    ):
+        bath = DrudeLorentzBath(
+            Q=coup_op,
+            lam=coup_strength,
+            gamma=cut_freq,
+            Nk=N_exp - 1,
+            T=temperature,
+            terminator=bnd_cut_approx,
+        )
+
+        if bnd_cut_approx:
+            # upgrade H_sys to a Liouvillian if needed and add the
+            # bath terminator
+            H_sys = self._convert_h_sys(H_sys)
+            is_timedep = isinstance(H_sys, QobjEvo)
+            H0 = H_sys.to_list()[0] if is_timedep else H_sys
+            is_hamiltonian = H0.type == "oper"
+            if is_hamiltonian:
+                H_sys = liouvillian(H_sys)
+            H_sys = H_sys + bath.terminator
+
+        super().__init__(H_sys, bath=bath, max_depth=N_cut, options=options)
+
+        # store input parameters as attributes for politeness and compatibility
+        # with HSolverDL in QuTiP 4.6 and below.
+        self.coup_strength = coup_strength
+        self.cut_freq = cut_freq
+        self.temperature = temperature
+        self.N_exp = N_exp
+        self.bnd_cut_approx = bnd_cut_approx
+
+
+class FermionicHEOMSolver(HEOMSolver):
+    """
+    A helper class for creating an :class:`HEOMSolver` for a single
+    fermionic bath.
+
+    See :class:`HEOMSolver` and :class:`FermionicBath` for more detailed
+    descriptions of the solver and bath parameters.
+
+    Attributes
+    ----------
+    H_sys : Qobj or QobjEvo or list
+        The system Hamiltonian or Liouvillian. See :class:`HEOMSolver` for
+        a complete description.
+
+    Q : Qobj
+        Operator describing the coupling between system and bath.
+        See :class:`FermionicBath` for a complete description.
+
+    ck_plus, vk_plus, ck_minus, vk_minus : lists
+        Lists containing coefficients of the fitted bath correlation
+        functions. See :class:`FermionicBath` for a complete description.
+
+    max_depth : int
+        The maximum depth of the heirarchy (i.e. the maximum number of bath
+        exponent "excitations" to retain). See :class:`HEOMSolver` for
+        a complete description.
+
+    options : :class:`qutip.solver.Options`
+        Generic solver options. If set to None the default options will be
+        used. See :class:`HEOMSolver` for a complete description.
+    """
+    def __init__(
+        self, H_sys, Q, ck_plus, vk_plus, ck_minus, vk_minus, max_depth,
+        options=None,
+    ):
+        bath = FermionicBath(Q, ck_plus, vk_plus, ck_minus, vk_minus)
+        super().__init__(
+            H_sys, bath=bath, max_depth=max_depth, options=options
+        )
+
+
+class _GatherHEOMRHS:
```

Comment:
Just a not for posterity, we did some checking of @hodgestar's new gather method against the direct inserts in CSR matrices, the old cython pad_csr method, and against tarun's c++ code.   This gather method definitely beats both the old python/cython methods, and seems to scale much better as we increase the problem size. C++ still wins overall, but this makes the python construction of the RHS much more useable in big problems.